dex/runner.go

```diff
@@ -86,38 +87,94 @@ type Connector interface {
+	mtx  sync.RWMutex
+	done chan struct{}
 }
 
-// NewConnectionMaster is the constructor for a new ConnectionMaster.
+// NewConnectionMaster creates a new ConnectionMaster. The Connect method should
+// be used before Disconnect. The On, Done, and Wait methods may be used at any
+// time. However, prior to Connect, Wait and Done immediately return and signal
+// completion, respectively.
 func NewConnectionMaster(c Connector) *ConnectionMaster {
 	return &ConnectionMaster{
 		connector: c,
 	}
 }
 
 // Connect connects the Connector, and returns any initial connection error. Use
-// Disconnect to shut down the Connector.
+// Disconnect to shut down the Connector. Connect is not safe for concurrent use
+// with Disconnect, and must be used before Disconnect.
 func (c *ConnectionMaster) Connect(ctx context.Context) (err error) {
-	c.init(ctx)
+	if c.On() { // probably a bug in the consumer
+		return errors.New("already running")
+	}
+
+	// Attempt to start the Connector.
+	ctx, cancel := context.WithCancel(ctx)
+	wg, err := c.connector.Connect(ctx)
+	if err != nil {
+		cancel() // no context leak
+		return err
+	}
+
+	// It's running, enable Disconnect.
+	c.cancel = cancel // caller should synchronize Connect/Disconnect calls
+
+	// Done and On may be checked at any time.
+	done := make(chan struct{})
 	c.mtx.Lock()
-	c.wg, err = c.connector.Connect(c.ctx)
+	c.done = done // after setting cancel
 	c.mtx.Unlock()
-	return err
+
+	go func() { // capture the local variables
+		wg.Wait()
+		cancel() // if the Connector just died on its own, don't leak the context
+		close(done)
+	}()
+
+	return nil
 }
 
-// Disconnect closes the connection and waits for shutdown.
-func (c *ConnectionMaster) Disconnect() {
-	c.cancel()
+// Done returns a channel that is closed when the Connector's WaitGroup is done.
+// If called before Connect, a closed channel is returned.
+func (c *ConnectionMaster) Done() <-chan struct{} {
 	c.mtx.RLock()
-	c.wg.Wait()
-	c.mtx.RUnlock()
+	defer c.mtx.RUnlock()
+	if c.done == nil { // before first Connect
+		done := make(chan struct{})
+		close(done)
+		return done
+	}
+	return c.done
+}
+
+// On indicates if the Connector is running. This returns false if never
+// connected, or if the Connector has completed shut down.
+func (c *ConnectionMaster) On() bool {
+	select {
+	case <-c.Done():
+		return false
+	default:
+		return true
+	}
 }
 
-// Wait waits for the the WaitGroup returned by Connect.
+// Wait waits for the the Connector to shut down. It returns immediately if
+// Connect has not been called yet.
 func (c *ConnectionMaster) Wait() {
-	c.wg.Wait()
-	c.cancel() // if not called from Disconnect, would leak context
+	<-c.Done() // let the anon goroutine from Connect return
+}
+
+// Disconnect closes the connection and waits for shutdown. This must not be
+// used before or concurrently with Connect.
+func (c *ConnectionMaster) Disconnect() {
+	if !c.On() {
+		return // probably a bug in the consumer
+	}
+	c.cancel()
```

Comment:
See the old `Disconnect` method where the `cancel` field has always been ungaurded by the mutex.
Evidently we do have sane Connect/Disconnect access patterns, and the `On` method is the only one that really needs concurrency controls, which it gets via `Done`.

Reply:
> This must not be used before or concurrently with Connect.

I think this is an assumption we are under at various levels. Good to doc it, though.